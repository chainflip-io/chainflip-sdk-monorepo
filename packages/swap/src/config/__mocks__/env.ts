export default {
  RPC_NODE_HTTP_URL: 'http://rpc-node.test',
  START_HTTP_SERVICE: false,
  START_PROCESSOR: false,
  SWAPPING_APP_PORT: 8080,
  BITCOIN_DEPOSIT_TRACKER_URL: '',
  RPC_BROKER_HTTPS_URL: 'https://rpc-broker.test',
  RPC_NODE_WSS_URL: 'ws://rpc-node.test',
  CHAINFLIP_NETWORK: 'perseverance',
  QUOTE_TIMEOUT: 1000,
  NODE_ENV: 'test',
  CHAINALYSIS_API_KEY: undefined,
  INGEST_GATEWAY_URL: 'https://ingest-gateway.test',
  PROCESSOR_BATCH_SIZE: 50,
  PROCESSOR_TRANSACTION_TIMEOUT: 10_000,
  REDIS_URL: 'redis://localhost:6379',
  MAINTENANCE_MODE: false,
  FULLY_DISABLED_INTERNAL_ASSETS: new Set(),
  DISABLED_DEPOSIT_INTERNAL_ASSETS: new Set(),
  DISABLED_DESTINATION_INTERNAL_ASSETS: new Set(),
  MAX_CHANNELS_OPEN_PER_ADDRESS: 10,
  DISABLE_QUOTING: false,
  SOLANA_RPC_HTTP_URL: 'http://solana-rpc.test',
  BITCOIN_RPC_HTTP_URL: 'http://bitcoin-rpc.test',
  STABLE_COIN_SLIPPAGE_MIN_PRICE: 0.995,
  DCA_CHUNK_INTERVAL_BLOCKS: 2,
  QUOTER_BALANCE_TOLERANCE_PERCENT: 10,
  QUOTING_BASE_SLIPPAGE: {},
  QUOTER_USE_AUGMENT: true,
  BROKER_COMMISSION_BPS: 0,
  ON_CHAIN_DEFAULT_NETWORK_FEE_BPS: 5n,
  ON_CHAIN_STABLECOIN_NETWORK_FEE_BPS: 1n,
  MINIMUM_NETWORK_FEE_USDC: 500_000n,
};
