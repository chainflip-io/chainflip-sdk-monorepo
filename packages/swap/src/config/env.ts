import { z } from 'zod';

const httpUrl = z.string().url().startsWith('http');
const wsUrl = z.string().url().startsWith('ws');

const optionalBoolean = z
  .string()
  .optional()
  .transform((value) => value?.toUpperCase() === 'TRUE');

const optionalNumber = (defaultValue: number) =>
  z
    .string()
    .optional()
    .transform((n) => Number(n) || defaultValue);

const chainflipNetwork = z.enum([
  'backspin',
  'sisyphos',
  'perseverance',
  'mainnet',
]);

const nodeEnv = z.enum(['development', 'production', 'test']);

export default z
  .object({
    RPC_NODE_HTTP_URL: httpUrl,
    START_HTTP_SERVICE: optionalBoolean,
    START_PROCESSOR: optionalBoolean,
    SWAPPING_APP_PORT: optionalNumber(8080),
    BITCOIN_DEPOSIT_TRACKER_URL: httpUrl.optional(),
    RPC_BROKER_HTTPS_URL: httpUrl,
    RPC_NODE_WSS_URL: wsUrl,
    CHAINFLIP_NETWORK: chainflipNetwork,
    QUOTE_TIMEOUT: optionalNumber(1000),
    NODE_ENV: nodeEnv.default('production'),
    CHAINALYSIS_API_KEY: z.string().optional(),
    INGEST_GATEWAY_URL: httpUrl,
    PROCESSOR_BATCH_SIZE: optionalNumber(50),
    PROCESSOR_TRANSACTION_TIMEOUT: optionalNumber(10_000),
  })
  // eslint-disable-next-line n/no-process-env
  .parse(process.env);
