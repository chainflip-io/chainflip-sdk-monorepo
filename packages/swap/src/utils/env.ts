import { ChainflipAsset } from '@chainflip/utils/chainflip';
import ServiceError from './ServiceError.js';
import env from '../config/env.js';

export const isLocalnet = () => env.CHAINFLIP_NETWORK === 'backspin';

export const assertRouteEnabled = ({
  srcAsset,
  destAsset,
}: {
  srcAsset: ChainflipAsset;
  destAsset: ChainflipAsset;
}) => {
  if (
    env.FULLY_DISABLED_INTERNAL_ASSETS.has(srcAsset) ||
    env.DISABLED_DEPOSIT_INTERNAL_ASSETS.has(srcAsset)
  ) {
    throw ServiceError.unavailable(`Asset ${srcAsset} is disabled`);
  }

  if (
    env.FULLY_DISABLED_INTERNAL_ASSETS.has(destAsset) ||
    env.DISABLED_DESTINATION_INTERNAL_ASSETS.has(destAsset)
  ) {
    throw ServiceError.unavailable(`Asset ${destAsset} is disabled`);
  }
};
