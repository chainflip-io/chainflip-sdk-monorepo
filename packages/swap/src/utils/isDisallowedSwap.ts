import { isAxiosError } from 'axios';
import { AML } from 'elliptic-sdk';
import { inspect } from 'util';
import { z } from 'zod';
import logger from './logger.js';
import prisma from '../client.js';
import env from '../config/env.js';

let elliptic: (typeof AML.prototype)['client'] | undefined;

const exposureSchema = z.object({ risk_score: z.number().nullable() });

const isTooRisky = async (address: string | undefined): Promise<boolean> => {
  if (!address || !env.ELLIPTIC_API_KEY || !env.ELLIPTIC_API_SECRET) {
    return false;
  }

  try {
    elliptic ??= new AML({ secret: env.ELLIPTIC_API_SECRET, key: env.ELLIPTIC_API_KEY }).client;

    const { data: result } = await elliptic.post('/v2/wallet/synchronous', {
      subject: {
        asset: 'holistic',
        blockchain: 'holistic',
        hash: address,
        type: 'address',
      },
      type: 'wallet_exposure',
    });

    const parsed = exposureSchema.parse(result);

    return parsed.risk_score !== null && parsed.risk_score >= env.ELLIPTIC_RISK_SCORE_TOLERANCE;
  } catch (error) {
    const level = isAxiosError(error) && error.status === 404 ? 'warn' : 'error';
    logger[level]('failed to request risk score from elliptic', { error: inspect(error) });
    return false;
  }
};

const isOnBlocklist = async (address: string | undefined): Promise<boolean> => {
  if (address === undefined) return false;

  const result = await prisma.blockedAddress.findFirst({
    where: {
      address: {
        equals: address,
        mode: 'insensitive',
      },
    },
  });

  return result !== null;
};

export default async function isDisallowedSwap(
  destAddress: string,
  srcAddress: string | undefined,
  refundAddress: string | undefined,
): Promise<boolean> {
  return (
    await Promise.all([
      isOnBlocklist(destAddress),
      isOnBlocklist(srcAddress),
      isOnBlocklist(refundAddress),
      isTooRisky(destAddress),
      isTooRisky(srcAddress),
      isTooRisky(refundAddress),
    ])
  ).some(Boolean);
}
