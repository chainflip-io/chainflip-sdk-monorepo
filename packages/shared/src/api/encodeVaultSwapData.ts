import { bytesToHex } from '@chainflip/utils/bytes';
import { chainflipNetworks } from '@chainflip/utils/chainflip';
import { priceX128ToPrice } from '@chainflip/utils/tickMath';
import { BigNumber } from 'bignumber.js';
import { z } from 'zod';
import { assert } from '../guards.js';
import { assetAndChain, chainflipAddress, hexString, unsignedInteger } from '../parsers.js';
import {
  affiliateBroker,
  ccmParamsSchema,
  dcaParams as dcaParamsSchema,
  fillOrKillParams as fillOrKillParamsSchema,
} from '../schemas.js';
import { validateAddress } from '../validation/addressValidation.js';

const TransformedCcmParams = ccmParamsSchema.transform(
  ({ message, gasBudget, ccmAdditionalData }) => ({
    message,
    gas_budget: gasBudget,
    ccm_additional_data: ccmAdditionalData,
    /** @deprecated DEPRECATED(1.10) we still need to pass cf_parameters until 1.8 is deployed to all networks */
    cf_parameters: ccmAdditionalData,
  }),
);

const FillOrKillParams = fillOrKillParamsSchema.transform(
  ({ retryDurationBlocks, refundAddress, minPriceX128, maxOraclePriceSlippage }) => ({
    retry_duration: retryDurationBlocks,
    refund_address: refundAddress!,
    min_price: `0x${BigInt(minPriceX128).toString(16)}` as const,
    max_oracle_price_slippage: maxOraclePriceSlippage,
  }),
);

const TransformedDcaParams = dcaParamsSchema.transform(
  ({ numberOfChunks, chunkIntervalBlocks }) => ({
    number_of_chunks: numberOfChunks,
    chunk_interval: chunkIntervalBlocks,
  }),
);

export const EncodeVaultSwapBody = z
  .object({
    srcAsset: assetAndChain,
    srcAddress: z.string().optional(),
    destAsset: assetAndChain,
    destAddress: z.string(),
    amount: unsignedInteger,
    commissionBps: z.number().optional().default(0),
    ccmParams: TransformedCcmParams.optional(),
    maxBoostFeeBps: z.number().optional(),
    affiliates: z.array(affiliateBroker).optional(),
    fillOrKillParams: FillOrKillParams,
    dcaParams: TransformedDcaParams.optional(),
    extraParams: z.object({ seed: hexString.optional() }).optional(),
    network: z.enum(chainflipNetworks),
    brokerAccount: chainflipAddress.optional(),
  })
  .superRefine(({ network, ...val }, ctx) => {
    if (val.srcAddress && !validateAddress(val.srcAsset.chain, val.srcAddress, network)) {
      ctx.addIssue({
        message: `Address "${val.srcAddress}" is not a valid "${val.srcAsset.chain}" address for "${network}"`,
        code: z.ZodIssueCode.custom,
      });
    }

    if (!validateAddress(val.destAsset.chain, val.destAddress, network)) {
      ctx.addIssue({
        message: `Address "${val.destAddress}" is not a valid "${val.destAsset.chain}" address for "${network}"`,
        code: z.ZodIssueCode.custom,
      });
    }

    if (!validateAddress(val.srcAsset.chain, val.fillOrKillParams.refund_address, network)) {
      ctx.addIssue({
        message: `Address "${val.fillOrKillParams.refund_address}" is not a valid "${val.srcAsset.chain}" address for "${network}"`,
        code: z.ZodIssueCode.custom,
      });
    }

    if (val.extraParams?.seed && val.extraParams.seed.length !== 66) {
      ctx.addIssue({ message: 'Seed must be 32 bytes', code: z.ZodIssueCode.custom });
    }
  })
  .transform(({ network, ...data }) => {
    let extraParams;
    if (data.srcAsset.chain === 'Bitcoin') {
      const minOutputAmount = BigNumber(data.amount.toString())
        .multipliedBy(priceX128ToPrice(data.fillOrKillParams.min_price))
        .toFixed(0);

      extraParams = {
        chain: 'Bitcoin',
        min_output_amount: `0x${BigInt(minOutputAmount).toString(16)}`,
        retry_duration: data.fillOrKillParams.retry_duration,
        max_oracle_price_slippage: data.fillOrKillParams.max_oracle_price_slippage,
      } as const;
    } else if (data.srcAsset.chain === 'Ethereum' || data.srcAsset.chain === 'Arbitrum') {
      extraParams = {
        chain: data.srcAsset.chain,
        input_amount: `0x${data.amount.toString(16)}`,
        refund_parameters: data.fillOrKillParams,
      } as const;
    } else if (data.srcAsset.chain === 'Solana') {
      assert(data.srcAddress, 'srcAddress is required for Solana');

      extraParams = {
        chain: 'Solana',
        from: data.srcAddress,
        seed: data.extraParams?.seed ?? bytesToHex(crypto.getRandomValues(new Uint8Array(32))),
        input_amount: `0x${data.amount.toString(16)}`,
        refund_parameters: data.fillOrKillParams,
      } as const;
    } else {
      throw new Error(`parameter encoding is not supported for ${data.srcAsset.chain}`);
    }

    return { ...data, extraParams };
  });

export const EncodedVaultSwapData = z.discriminatedUnion('chain', [
  z.object({
    chain: z.literal('Bitcoin'),
    nulldataPayload: hexString,
    depositAddress: z.string(),
  }),
  z.object({
    chain: z.enum(['Ethereum', 'Arbitrum']),
    value: z.string(),
    to: hexString,
    calldata: hexString,
    sourceTokenAddress: hexString.optional(),
  }),
  z.object({
    chain: z.literal('Solana'),
    programId: z.string(),
    data: hexString,
    accounts: z.array(
      z.object({
        pubkey: z.string(),
        isSigner: z.boolean(),
        isWritable: z.boolean(),
      }),
    ),
  }),
]);
