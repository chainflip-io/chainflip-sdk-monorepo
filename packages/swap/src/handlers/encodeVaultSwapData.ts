import { HttpClient } from '@chainflip/rpc';
import { getInternalAsset } from '@chainflip/utils/chainflip';
import { z } from 'zod';
import { getParameterEncodingRequestSchema } from '@/shared/broker.js';
import { formatResponse } from '@/shared/objects.js';
import { chainflipAddress } from '@/shared/parsers.js';
import env from '../config/env.js';
import { assertRouteEnabled } from '../utils/env.js';
import logger from '../utils/logger.js';
import { validateSwapAmount } from '../utils/rpc.js';
import ServiceError from '../utils/ServiceError.js';

const brokerClient = new HttpClient(env.RPC_BROKER_HTTPS_URL);
const nodeClient = new HttpClient(env.RPC_NODE_HTTP_URL);

export const encodeVaultSwapDataSchema = getParameterEncodingRequestSchema(
  env.CHAINFLIP_NETWORK,
).and(
  z.object({
    brokerAccount: chainflipAddress.optional(),
    brokerCommissionBps: z.number().optional(), // sdk version 1.8.3 sends brokerCommissionBps instead of commissionBps
  }),
);

export const encodeVaultSwapData = async (input: z.output<typeof encodeVaultSwapDataSchema>) => {
  logger.info('Fetching vault swap data', input);

  const srcAsset = getInternalAsset(input.srcAsset);
  const destAsset = getInternalAsset(input.destAsset);
  assertRouteEnabled({ srcAsset, destAsset });

  const result = await validateSwapAmount(srcAsset, BigInt(input.amount));

  if (!result.success) throw ServiceError.badRequest(result.reason);

  const response = input.brokerAccount
    ? await nodeClient.sendRequest(
        'cf_request_swap_parameter_encoding',
        input.brokerAccount,
        input.srcAsset,
        input.destAsset,
        input.destAddress,
        input.commissionBps || input.brokerCommissionBps || 0,
        input.extraParams,
        input.ccmParams,
        input.maxBoostFeeBps,
        input.affiliates,
        input.dcaParams,
      )
    : await brokerClient.sendRequest(
        'broker_request_swap_parameter_encoding',
        input.srcAsset,
        input.destAsset,
        input.destAddress,
        0, // default broker account does not support commission
        input.extraParams,
        input.ccmParams,
        input.maxBoostFeeBps,
        undefined, // default broker account does not support affiliates
        input.dcaParams,
      );

  logger.info('Vault swap data fetched', response);

  return formatResponse(response);
};
