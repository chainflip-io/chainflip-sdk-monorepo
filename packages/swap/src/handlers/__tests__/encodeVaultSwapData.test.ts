import { vi, describe, it, beforeEach, afterEach, expect } from 'vitest';
import { environment, mockRpcResponse } from '@/shared/tests/fixtures';
import env from '@/swap/config/env';
import isDisallowedSwap from '../../utils/isDisallowedSwap';
import { encodeVaultSwapData } from '../encodeVaultSwapData';

vi.mock('@/shared/broker', async (importOriginal) => {
  const original = (await importOriginal()) as object;
  return {
    ...original,
    requestSwapParameterEncoding: vi.fn(),
  };
});

vi.mock('../../utils/isDisallowedSwap', () => ({
  default: vi.fn().mockResolvedValue(false),
}));

describe(encodeVaultSwapData, () => {
  let oldEnv: typeof env;

  beforeEach(async () => {
    oldEnv = structuredClone(env);
  });

  afterEach(() => {
    Object.assign(env, oldEnv);
  });

  it('calls the broker api with the correct params', async () => {
    const postSpy = mockRpcResponse((url, data: any) => {
      if (data.method === 'cf_environment') {
        return Promise.resolve({ data: environment({ maxSwapAmount: '0x10000000000' }) });
      }

      if (data.method === 'broker_request_swap_parameter_encoding') {
        return Promise.resolve({
          data: {
            id: '1',
            jsonrpc: '2.0',
            result: {
              chain: 'Bitcoin',
              nulldata_payload:
                '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
              deposit_address: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
            },
          },
        });
      }

      throw new Error(`unexpected axios call to ${url}: ${JSON.stringify(data)}`);
    });

    const result = await encodeVaultSwapData({
      srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
      destAsset: { asset: 'ETH', chain: 'Ethereum' },
      destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
      amount: 175000000n,
      commissionBps: 15,
      extraParams: {
        chain: 'Bitcoin',
        min_output_amount: '0xfc6f7c40458122964d0000000',
        retry_duration: 500,
      },
      fillOrKillParams: {
        retry_duration: 500,
        refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
        min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
      },
      ccmParams: {
        gas_budget: '0x75bcd15',
        message: '0xdeadc0de',
        cf_parameters: undefined,
        ccm_additional_data: undefined,
      },
      maxBoostFeeBps: 35,
      affiliates: [
        {
          account: 'cFHyJEHEQ1YkT9xuFnxnPWVkihpYEGjBg4WbF6vCPtSPQoE8n',
          bps: 10,
        },
      ],
      dcaParams: {
        number_of_chunks: 100,
        chunk_interval: 5,
      },
    });

    expect(postSpy).toHaveBeenCalledWith('https://rpc-broker.test', [
      {
        id: expect.any(String),
        jsonrpc: '2.0',
        method: 'broker_request_swap_parameter_encoding',
        params: [
          {
            asset: 'BTC',
            chain: 'Bitcoin',
          },
          {
            asset: 'ETH',
            chain: 'Ethereum',
          },
          '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
          0, // ignores commission
          {
            chain: 'Bitcoin',
            min_output_amount: '0xfc6f7c40458122964d0000000',
            retry_duration: 500,
          },
          {
            gas_budget: '0x75bcd15',
            message: '0xdeadc0de',
          },
          35,
          null, // ignores affiliates
          {
            chunk_interval: 5,
            number_of_chunks: 100,
          },
        ],
      },
    ]);
    expect(result).toEqual({
      chain: 'Bitcoin',
      nulldataPayload:
        '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
      depositAddress: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
    });
  });

  it('calls the node rpc with the correct params if broker account is given', async () => {
    const postSpy = mockRpcResponse((url, data: any) => {
      if (data.method === 'cf_environment') {
        return Promise.resolve({ data: environment({ maxSwapAmount: '0x10000000000' }) });
      }

      if (data.method === 'cf_request_swap_parameter_encoding') {
        return Promise.resolve({
          data: {
            id: '1',
            jsonrpc: '2.0',
            result: {
              chain: 'Bitcoin',
              nulldata_payload:
                '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
              deposit_address: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
            },
          },
        });
      }

      throw new Error(`unexpected axios call to ${url}: ${JSON.stringify(data)}`);
    });

    const result = await encodeVaultSwapData({
      srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
      destAsset: { asset: 'ETH', chain: 'Ethereum' },
      destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
      amount: 175000000n,
      brokerAccount: 'cFLdocJo3bjT7JbT7R46cA89QfvoitrKr9P3TsMcdkVWeeVLa',
      commissionBps: 15,
      extraParams: {
        chain: 'Bitcoin',
        min_output_amount: '0xfc6f7c40458122964d0000000',
        retry_duration: 500,
      },
      fillOrKillParams: {
        retry_duration: 500,
        refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
        min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
      },
      ccmParams: {
        gas_budget: '0x75bcd15',
        message: '0xdeadc0de',
        cf_parameters: undefined,
      },
      maxBoostFeeBps: 35,
      affiliates: [
        {
          account: 'cFHyJEHEQ1YkT9xuFnxnPWVkihpYEGjBg4WbF6vCPtSPQoE8n',
          bps: 10,
        },
      ],
      dcaParams: {
        number_of_chunks: 100,
        chunk_interval: 5,
      },
    });

    expect(postSpy).toHaveBeenCalledWith('http://rpc-node.test', [
      {
        id: expect.any(String),
        jsonrpc: '2.0',
        method: 'cf_request_swap_parameter_encoding',
        params: [
          'cFLdocJo3bjT7JbT7R46cA89QfvoitrKr9P3TsMcdkVWeeVLa',
          {
            asset: 'BTC',
            chain: 'Bitcoin',
          },
          {
            asset: 'ETH',
            chain: 'Ethereum',
          },
          '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
          15,
          {
            chain: 'Bitcoin',
            min_output_amount: '0xfc6f7c40458122964d0000000',
            retry_duration: 500,
          },
          {
            gas_budget: '0x75bcd15',
            message: '0xdeadc0de',
          },
          35,
          [
            {
              account: 'cFHyJEHEQ1YkT9xuFnxnPWVkihpYEGjBg4WbF6vCPtSPQoE8n',
              bps: 10,
            },
          ],
          {
            chunk_interval: 5,
            number_of_chunks: 100,
          },
        ],
      },
    ]);
    expect(result).toEqual({
      chain: 'Bitcoin',
      nulldataPayload:
        '0x0003656623d865425c0a4955ef7e7a39d09f58554d0800000000000000000000000000000000000001000200000100',
      depositAddress: 'bcrt1pmrhjpvq2w7cgesrcrvuhqw6n6j487l6uc7tmwtx9jen7ezesunhqllvzxx',
    });
  });

  it('rejects sanctioned addresses', async () => {
    vi.mocked(isDisallowedSwap).mockResolvedValueOnce(true);

    await expect(
      encodeVaultSwapData({
        srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
        destAsset: { asset: 'ETH', chain: 'Ethereum' },
        srcAddress: 'bc1qqwykx04uenc842d3sf50cjtehtj9tenugk808w',
        destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: 'bc1qv6nkwfd0l7gjucvccgd4k4qcea488cfkg7zjcp',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Failed to get vault swap data, please try again later');
    expect(vi.mocked(isDisallowedSwap).mock.calls).toMatchSnapshot();
  });

  it('rejects if source asset is disabled', async () => {
    env.DISABLED_INTERNAL_ASSETS = ['Flip', 'Btc'];

    await expect(
      encodeVaultSwapData({
        srcAsset: { asset: 'BTC', chain: 'Bitcoin' },
        destAsset: { asset: 'ETH', chain: 'Ethereum' },
        destAddress: '0xe983fD1798689eee00c0Fb77e79B8f372DF41060',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Asset Btc is disabled');
  });

  it('rejects if destination asset is disabled', async () => {
    env.DISABLED_INTERNAL_ASSETS = ['Btc', 'Dot'];

    await expect(
      encodeVaultSwapData({
        srcAsset: { asset: 'ETH', chain: 'Ethereum' },
        destAsset: { asset: 'DOT', chain: 'Polkadot' },
        destAddress: '1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo',
        amount: 175000000n,
        commissionBps: 15,
        extraParams: {
          chain: 'Bitcoin',
          min_output_amount: '0xfc6f7c40458122964d0000000',
          retry_duration: 500,
        },
        fillOrKillParams: {
          retry_duration: 500,
          refund_address: '0xa56A6be23b6Cf39D9448FF6e897C29c41c8fbDFF',
          min_price: '0x152d02c7e14af680000000000000000000000000000000000000',
        },
      }),
    ).rejects.toThrow('Asset Dot is disabled');
  });
});
