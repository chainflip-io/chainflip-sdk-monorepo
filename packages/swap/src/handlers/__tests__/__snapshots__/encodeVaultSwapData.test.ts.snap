// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encodeVaultSwapData > rejects sanctioned addresses 1`] = `
[
  [
    "0xe983fD1798689eee00c0Fb77e79B8f372DF41060",
    "bc1qqwykx04uenc842d3sf50cjtehtj9tenugk808w",
    "bc1qv6nkwfd0l7gjucvccgd4k4qcea488cfkg7zjcp",
  ],
]
`;
